### ISSUE TITLE
The uc_reg_read/write API is slow.

### ISSUE THREAD

**Issue by farmdve** [unicorn-engine/unicorn#318]
I've compiled Unicorn with -Ofast and my own code with -Ofast as well. I made speed improvements in both these functions.

So I did a small benchmark. I ran Unicorn on a real-world DRM and executed exactly 1,047,269,679 instructions in ~15 seconds. I executed this code with a UC_HOOK_CODE hook that counted the instructions.
If in my hook I add just a single uc_reg_read or uc_reg_write, the execution of the same instructions took 21 seconds, thats 6-7 seconds more with just one uc_reg_write per instruction. If I read all register, execution took ~70 seconds. My speed improvements are that I removed `X86_CPU(uc, mycpu)->env.regs` from each register in the switch table, `X86_CPU` calls lots of object checking code and this caused an even more severe penalty.

So I did this

``` c
    CPUState *mycpu = first_cpu;
    CPUX86State *env = mycpu->env_ptr;
```

then I simply do `env->regs[R_EAX] = *(uint32_t *)value;` this shaved off a few seconds of execution for me.

But overall uc_reg_read and uc_reg_write are slow. I was thinking of how to speed them up and my only idea was to use a structure array of pointers to each register and then just access them via `regid`, but my implementation was wonky, moreover some registers were in env->segs, others e.g in env->eip.
Furthermore in uc_reg_read or write we shouldn't check for the mode(e.g if it's 16-bit or 32-bit) we should have separate functions for those, as not to incur a penalty(however small it may be).

**Comment by lunixbochs** [comment]
#8

**Comment by farmdve** [comment]
@lunixbochs

So, let's start the ball rolling. Let's optimise the API(by designing a new one or optimising this one) and then start tweaking the internals of QEMU for more speed.

**Comment by lunixbochs** [comment]
I propose something like this:

    uc_err uc_reg_read_all(handle, uint64_t *out) {
        CPUState *mycpu = X86_CPU_FAST(handle);
        CPUX86State *env = mycpu->env_ptr;
        for (int i = 0; i < UC_X86_ALL_REG_COUNT) {
            out[i] = reg_get_macro(env, i);
        }
        return UC_ERR_OK;
    }

and a constant to determine the register count for allocations:

    UC_X86_ALL_REG_COUNT

**Comment by farmdve** [comment]
The register enum is quite large. I think we need to separate the API so we have different functions for GPRs and special registers.

Or if not, I was thinking of some way to have a branchless copy of the register. I was thinking of a structure array filled with pointers to each register.

Then in x86_reg_read/write instead of having a switch table we'd have `*regArr[regid] = *value` or `*value = *regArr[regid]`

The problem is that the API separates each register to it's smaller counterparts(AX,AL,AH) and we need special checks to write the correct size.

**Comment by JonathonReinhart** [comment]
@farmdve This was discussed in #8 and never really went anywhere. I proposed a `union` for elegantly handling the various-sized x86 registers.  I'm not convinced it's the best method though.

**Comment by aquynh** [comment]
yes it went nowhere since nobody stepped up for it. while this API is good to have, perhaps the interest is not big enough?

so if anybody really wants this API, go right ahead doing it.

**Comment by farmdve** [comment]
@aquynh 

We need to first decide how to do it.

**Comment by aquynh** [comment]
to keep it simple, this API should work with full registers, but not subregisters (such as AL, AX)

**Comment by JonathonReinhart** [comment]
> to keep it simple, this API should work with full registers, but not subregisters (such as AL, AX)

@aquynh That's easier said than done. What is a "full register"? Do we promote everything to 512 bits because that's the size of an AVX2 register?

**Comment by lunixbochs** [comment]
No, I don't think floating point or SIMD registers should be included by default. If anyone needs that, I think a "grab special registers" API should exist that fills byte buffers instead of uint64... because SIMD and floating point registers aren't guaranteed to be directly representable in any native type.

**Comment by lunixbochs** [comment]
Also, I still think some of the problem is the stack/function call overhead from calling back into Unicorn to get registers. That's why my original proposal was to maintain a list of pointers into the CPU's register struct, and just pass the whole list for free into hook functions.

**Comment by JonathonReinhart** [comment]
I agree with @lunixbochs. What would an emulator do in a hook without this context? Very little. So the context will, more likely than not, be desired by the hooks.

**Comment by farmdve** [comment]
@lunixbochs
You are correct, the core problem is this currently. Some instruction is jitted, since we've hooked code, almost every single time or every time the eflags are synced, that's the first, then our callback is called, but first the EIP has to be synced. Then our actual hook is called, obviously as you know calling a function passed parameters to the stack via push, inside a function the stack frame is saved, stack might be reserved. If we call uc_reg_read/write, thats the same procedure again, push arguments to stack, enter function, reserve stack, branch on uc mode(e.g if 16,32,64-bit), then branch on register, and it checks to see which regid matches with the register we want. Then, the most expensive of calls is X86_CPU macro which calls various object checking functions.

**Comment by lunixbochs** [comment]
@farmdve can you run this test again using the `reg_read_batch` API from #506?

**Comment by farmdve** [comment]
This can be further improved by storing a static pointer to the env
pointer,  because calling X86_CPU() or ARM_CPU casts are quite expensive.
Removing those will further improve performance.

**Comment by lunixbochs** [comment]
removing X86_CPU conversion should be possible with https://github.com/unicorn-engine/unicorn/issues/636#issuecomment-249103954

**Comment by lunixbochs** [comment]
this should be a bit better with reg_read/write_batch and #815

**Comment by farmdve** [comment]
Like I said before, when reading a single register, Unicorn is doing lots
of calls for type conversion, e.g X86_CPU(uc, mycpu)->env, this is
completely redundant, and having a static variable in my opinion holding a
(static)reference to X86_CPU(uc, mycpu)->env would reduce the call overhead
by even further.

**Comment by lunixbochs** [comment]
I fixed X86_CPU in #815. These casts no longer happen on any arch.

**Comment by JonathonReinhart** [comment]
@farmdve A `static` variable prevents one from using multiple instances of unicorn at the same time.
